### PR TITLE
example of a user defined type

### PR DESCRIPTION
**Pull request by jkozdon**

### CODE REVIEW

**Comment by jkozdon** [comment]
There doesn't seem to be an example showing what's required for the user to define their own type.

File: examples/User_Type.jl [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+typealias DataFloat Float64
+
+type Vec2
+  a::DataFloat
+  b::DataFloat
+end
+
+# Necessary for the solver
+Base.(:/)(x::Vec2, y::Real) = Vec2(x.a/y, x.b/y)
```
Comment (mauro3) [inline examples/User_Type.jl]: To make this 0.5-proof: `@compat Base.:+`

File: examples/User_Type.jl [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+Base.(:*)(y::Real, x::Vec2) = Vec2(y * x.a, y * x.b)
+Base.(:*)(x::Vec2, y::Real) = y*x
+Base.(:.*)(y::Real, x::Vec2) = y*x
+Base.(:+)(x::Vec2, y::Vec2) = Vec2(x.a + y.a, x.b + y.b)
+Base.(:-)(x::Vec2, y::Vec2) = Vec2(x.a - y.a, x.b - y.b)
+Base.norm(x::Vec2) = sqrt(x.a^2 + x.b^2)
+Base.zero(x::Type{Vec2}) = Vec2(zero(DataFloat), zero(DataFloat))
+Base.isnan(x::Vec2) = isnan(x.a) || isnan(x.b)
```
Comment (mauro3) [inline examples/User_Type.jl]: This should be `ODE.isoutofdomain`

File: examples/User_Type.jl [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+function f(t, y)
+  (a, b) = (y.a, y.b)
+  Vec2(-b, a)
+end
+
+let
```
Comment (mauro3) [inline examples/User_Type.jl]: I'd leave the `let` block away.

File: examples/User_Type.jl [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  # Initial condtions
+  const start = Vec2(0.0, 0.1)
```
Comment (mauro3) [inline examples/User_Type.jl]: These don't need to be `const` for performance as they are passed into the function as arguments.

**Comment by mauro3** [comment]
Thanks, this is good!  Maybe a few sentences comments would be good: what the example is about, maybe a word about the ODE.

There is also the test-case `test/interface-tests.jl`.  You could mention it in your example.

**Comment by mauro3** [comment]
Looks good to me.  Someone please merge this, unless there are objections.

**Comment by ChrisRackauckas** [comment]
Looks good to me as well. Merging.

**Comment by mauro3** [comment]
Ah, yes, we can merge now! ;-)  Thanks!

**Pull request merged**